### ISSUE TITLE
VLM: Partitioning for creating RDD from Seq[RasterReader]

### ISSUE THREAD

**Issue by echeipesh** [geotrellis/geotrellis-contrib#6]
When we are creating `RDD` we have a set of GeoTiffs and corresponding `RasterReader`s that we are reading to match a given `LayoutDefinition`. The resampling and reprojection of for each key happens during read time. 

However we can impose no restriction on relative position of the input rasters so the following cases are possible:

1. None of the input rasters overlap
2. Some input rasters partially overlap
3. Input rasters heavily overlap

An explicit requirement is to avoid shuffling raster tiles after they have been read.
In other words for every `SpatialKey` we need to read we must sample all input rasters that intersect it.

Mostly the implication that the strategy used in `GeoTiffRDD` is _not_ appropriate here. Which is:
1. Read metadata for each GeoTiff into single partition
2. Partition resulting tiles from each GeoTiff and repartition to increase the partition count dedicated to a large file
3. Read (GeoTiff, window) pairs to produce tiles

**Comment by echeipesh** [comment]
One way to look at this is say that we're partitioning space covered by input rasters rather than input rasters themselves. For instance we can use a quad tree to continue partitioning space, refining quads until they contain a small enough set of tiles and discarding quads that do not intersect any input rasters.

Another more direct way is to flatMap each `RasterReader` into set of keys that it will need to read , producing records of shape `(SpatialKey, RasterReader)`, and joining on key using arbitrary partitioner.

**Comment by echeipesh** [comment]
Superseded by https://github.com/geotrellis/geotrellis-contrib/issues/180

Also `RasterReader` was a sketch name for `RasterSource` for context.
